Revised to support `DeferredChannels` and a `@defer` macro for `Remot…

rofinn opened the pull request:
…eChannel` and `Future`.

NOTES:
- Tried including a `serialize` method to better fit the `Future` and `RemoteChannel` interfaces, but this caused tests to stall at `remotecall_wait`.
- It would be nice if we could combine `DeferredChannel` and `DeferredFuture` into a common wrapper type that dispatched on what was being wrapped.

rofinn committed:
Revised to support `DeferredChannels` and a `@defer` macro for `RemoteChannel` and `Future`.

NOTES:
- Tried including a `serialize` method to better fit the `Future` and `RemoteChannel` interfaces, but this caused tests to stall at `remotecall_wait`.
- It would be nice if we could combine `DeferredChannel` and `DeferredFuture` into a common wrapper type that dispatched on what was being wrapped.

iamed2 commented on src/DeferredFutures.jl:
```diff
@@ -1,35 +1,88 @@
+    elseif ex.args[1] == :RemoteChannel
+        return Expr(:call, :DeferredChannel, ex.args[2:end]...)
+    else
+        return parse("throw(AssertionError(\"Expected RemoteChannel or Future and got $(ex.args[1])\"))")
```
This error should just be thrown

rofinn replied:
I disagree because this makes testing that it throws the AssertionError difficult. A try catch doesn't seem to catch the AssertionError if it is thrown inside the macro.

iamed2 replied:
That's why you use `macroexpand` to test macros (you can use `@test_throws` there)

rofinn replied:
Ah, ok. Thanks, that comment makes more sense now.

iamed2 commented on src/DeferredFutures.jl:
```diff
@@ -1,35 +1,88 @@
+
+macro defer(ex::Expr)
+    @assert ex.head == :call
+    @assert isa(ex.args[1], Symbol)
```
These asserts should throw errors like the one below

iamed2 commented on src/DeferredFutures.jl:
```diff
@@ -1,35 +1,88 @@
+        put!(ref.outer, inner)
+    end
+
+    # Feature the `RemoteChannel` and `put!`
```
typo

iamed2 commented on src/DeferredFutures.jl:
```diff
@@ -1,35 +1,88 @@
+
+function Base.put!(ref::DeferredChannel, val)
+    # On the first call to put! create the `RemoteChannel`
+    # and `put!`` it in the `Future`
```
typo (extra backtick)

iamed2 commented on src/DeferredFutures.jl:
```diff
@@ -1,35 +1,88 @@
+    func::Function      # A function the generates the RemoteChannel
+end
+
+function DeferredChannel(f::Function, pid::Integer=myid())
```
DeferredFutures are API-compatible with Futures. To be API-compatible with RemoteChannels, the function accepted should be the one which creates the inner channel, not the one which creates the RemoteChannel.

iamed2 commented on test/runtests.jl:
```diff
@@ -107,3 +145,27 @@ end
     end
 end
+
+@testset "@defer" begin
```
Test macros using `macroexpand`

iamed2 commented:
Also, add to the README please :)

rofinn committed:
Minor changes to DeferredChannel, @defer and test cases.

- DeferredChannel.func is a Channel (not RemoteChannel) generating function
- @defer uses a raw throw(AsssertionError()) everywhere
- @defer test cases fixed to use macroexpand

rofinn committed:
Updated README with `DeferredChannel` and `@defer` usage.

iamed2 commented on README.md:
````diff
@@ -5,15 +5,44 @@
+julia> DeferredChannel(4, 128; content=Int)
+DeferredFutures.DeferredChannel(Future(4,1,2,Nullable{Any}()),DeferredFutures.#2)
+```
+Note that `DeferredChannel()` will create a `RemoteChannel` with `RemoteChannel(()->Channel{Any}(32), myid())` by default.
````
The default channel for `RemoteChannel` is `Channel{Any}(1)`

rofinn replied:
Yeah, but this always seemed like a bad API choice given that by default `Channel()` returns `Channel{Any}(32)`. Since it seems odd you'd want to create a `RemoteChannel` of size 1, I opted for the behaviour of regular `Channel`s.

iamed2 replied:
As of 1.0 `Channel()` will be removed and channels will not have a default size. `RemoteChannel()` has a substantially different code path than when specifying a size, so it seems reasonable to keep it rather than preventing that from being usable.

iamed2 commented on test/runtests.jl:
```diff
@@ -107,3 +145,36 @@ end
 end
+
+@testset "@defer" begin
+    channel = eval(macroexpand(quote
```
Both tests with `@defer` should exist as well as tests with `macroexpand`. Neither should include `eval`. Just test that the returned code is correct and that no errors are thrown when running `macroexpand`. Also test for error conditions, which would have uncovered your "`AssertertionError`".

iamed2 reviewed (changes requested):
If you plan on using `DeferredChannel` for Dispatcher, duplicate the "Allocation" and "Transfer" tests for `DeferredChannel`.

rofinn committed:
Changed default size of `DeferredChannel` and fixed `@defer` tests to avoid `eval`.

Commit:
Update README.md
